### ISSUE TITLE
Greenlets not getting cleaned up

### ISSUE THREAD

**Issue by mikeroup** [VOLTTRON/volttron#765]
I'm mostly interested in ones that have dictionaries like the following. These can be easily created by `volttron-ctl status`. There are greelets that look like this before control is ever called which makes me think it may be an agent/core issue and not a control issue.

```
{
    '_notifier': <callback at 0x7fba7038e990 stopped>,
    '_start_event': <gevent.greenlet._dummy_event object at 0x7fba75ad9f50>,
    'value': None,
    '_links': deque([]),
    '_exc_info': (None, None, None)
}
```

I have a service like the one below to look at greenlets over time. The number of greenlets only increases. I used print instead of the logging module so that I can run `$ volttron` and have only this output.
``` python
import gc
from greenlet import greenlet
import pdb

class MemoryAnalysis(Agent):
    def __init__(self, *args, **kwargs):
        super(MemoryAnalysis, self).__init__(*args, **kwargs)
        self.iteration = 0
        self.oldstate = set(self.get_greenlets())

    @Core.periodic(15)
    def show_usage(self):
        print("================================================================================")
        print("ITERATION {}".format(self.iteration))
        self.iteration += 1

        # clean up unneeded objects
        gc.collect()

        # get all greenlets
        newstate = set(self.get_greenlets())
        print("THERE ARE {} GREENLETS".format(len(newstate)))

        # put previously untracked greenlets in diff
        diff = newstate - self.oldstate
        self.oldstate = newstate

        # pdb.set_trace()

    def get_greenlets(self):
        return [g for g in gc.get_objects() if isinstance(g, greenlet)]
```

**Comment by mikeroup** [comment]
After some more investigating I generated the following graph with the objgraph module. Most of it seems to be noise from the necessary local references and pdb. I highlighted an interesting part.

![outline](https://cloud.githubusercontent.com/assets/4081515/18819415/7b846c38-8345-11e6-8028-1118b02eae8a.png)

**Comment by mikeroup** [comment]
I used the inspect module to make greenlets record the stack when they are created. Looks like leftover greenlets are going through this execution pathway. 

```
(<frame object at 0x7f7bac10bbf0>, '/home/mike/volttron/env/local/lib/python2.7/site-packages/gevent/greenlet.py', 396, 'spawn_later', ['        g = cls(*args, **kwargs)\n'], 0)
(<frame object at 0x7f7bac109830>, '/home/mike/volttron/volttron/platform/vip/agent/core.py', 348, 'spawn_later', ['        greenlet = gevent.spawn_later(seconds, func, *args, **kwargs)\n'], 0)
(<frame object at 0x7f7bac0ed608>, '/home/mike/volttron/volttron/platform/vip/agent/subsystems/pubsub.py', 143, '_peer_add', ['        self.core().spawn_later(delay, self.synchronize, peer)\n'], 0)
(<frame object at 0x7f7bac0ed420>, '/home/mike/volttron/volttron/platform/vip/agent/dispatch.py', 82, 'send', ['                for receiver in self._receivers]\n'], 0)
(<frame object at 0x7f7bac109638>, '/home/mike/volttron/volttron/platform/vip/agent/subsystems/peerlist.py', 102, '_handle_subsystem', ["            getattr(self, 'on' + op).send(self, peer=peer)\n"], 0)
(<frame object at 0x17345f0>, '/home/mike/volttron/volttron/platform/vip/agent/core.py', 664, 'vip_loop', ['                    handle(message)\n'], 0)
(<frame object at 0x7f7bb406d230>, '/home/mike/volttron/env/local/lib/python2.7/site-packages/gevent/greenlet.py', 536, 'run', ['                result = self._run(*self.args, **self.kwargs)\n'], 0)
```

**Comment by mikeroup** [comment]
The links that we were using in core
```
self.greenlet.link(lambda glt: greenlet.kill())
```
were keeping finished greenlets from getting garbaged collected. Addressed by keeping greenlet references in a weakset instead of telling the core to kill each one individually.

**Comment by mikeroup** [comment]
Resolved in #869
